hack/windows/Install-OVS.ps1

```diff
@@ -244,49 +423,197 @@ function InstallDependency() {
+            return $ovsFile.FullName
+        }
+
+        # $ovsFile as a zip file is supported
+        $attributes = $ovsFile.Attributes
+        if (("$attributes" -eq "Archive") -and ($ovsFile.Extension -eq ".zip" ) ) {
+            $OVSZip = $LocalFile
+            $OVSDownloadURL = ""
+            $OVSPublishedHash = ""
+        } else {
+            Log "Unsupported local file $LocalFile; it should be a zip archive"
+            exit 1
+        }
+    } else {
+        $OVSZip = "$WorkDir\ovs-win64.zip"
+        if ($DownloadURL -ne "" -and $DownloadURL -ne "$OVSDownloadURL") {
+            $OVSDownloadURL = $DownloadURL
+            $desiredOVSPublishedHash = "invalid"
+        }
+    }
+
+    # Extract zip file to $OVSInstallDir
+    if (Test-Path -Path $OVSInstallDir) {
+        Log "$OVSInstallDir already exists, exit OVS installation."
+        exit 1
+    }
+    $removeZipFile = $false
+    if ($OVSDownloadURL -ne "") {
+        DownloadOVS -localZipFile $OVSZip -downloadURL $OVSDownloadURL -desiredHash $desiredOVSPublishedHash
+        $removeZipFile = $true
+    }
+    $ovsInstallParentPath = Split-Path -Path $OVSInstallDir -Parent
+    Expand-Archive -Path $OVSZip -DestinationPath $ovsInstallParentPath | Out-Null
+    if ($removeZipFile) {
+        rm $OVSZip
+    }
+    return $OVSInstallDir
+}
 
-DownloadOVS
+function CopyOVSUtilities() {
+    param (
+        [parameter(Mandatory = $true)] [string] $OVSInstallPath
+    )
+    $hostUsrBinPath = "C:\openvswitch\usr\bin"
+    $usrBinPath="${OVSInstallPath}\usr\bin"
+    if ("$usrBinPath" -ne "$hostUsrBinPath") {
+        if (-not $global:updateOVS) {
+            return
+        }
+        mkdir -p $hostUsrBinPath
+        cp -r $usrBinPath\* $hostUsrBinPath
+        AddToEnvPath($hostUsrBinPath)
+    }
+}
 
-InstallOVS
+function InstallOVS() {
+    param (
+        [parameter(Mandatory = $true)] [string] $OVSInstallPath
+    )
+    # Install powershell modules
+    $OVSScriptsPath = "${OVSInstallPath}\scripts"
+    CheckAndInstallScripts($OVSScriptsPath)
+
+    # Install VC redistributables.
+    $OVSRedistDir="${OVSInstallPath}\redist"
+    # Check if the VC redistributable is already installed. If not installed, or the installed version
+    # is lower than $MininalVCRedistVersion, install the local provided VC redistributable files or
+    # download the file and install it.
+    CheckAndInstallVCRedists -VCRedistPath $OVSRedistDir -VCRedistsVersion $MininalVCRedistVersion
+
+    # Install OVS driver.
+    $OVSDriverDir = "${OVSInstallPath}\driver"
+    Log "Installing OVS kernel driver"
+    CheckAndInstallOVSDriver($OVSDriverDir)
 
-if ($InstallUserspace -eq $true) {
-    InstallDependency
+    if ($InstallUserspace -eq $true) {
+        InstallOVSServices($OVSInstallPath)
+    } else {
+        # Copy OVS utilities from container path "openvswitch/usr/bin" to host path "c:/openvswitch/"
+        CopyOVSUtilities($OVSInstallPath)
+    }
```

Comment:
IMO, it is strange to only call this function when `$InstallUserspace` is `$false`, even though I understand why it is done this way. However, given that `CopyOVSUtilities` already checks `if ("$usrBinPath" -ne "$hostUsrBinPath")`, could we call this function unconditionally?
Right now, we are kind of treating `$InstallUserspace == $true` as "I am running in a container", which I think makes the script harder to understand. Renaming the function to `CopyOVSUtilitiesIfNeeded` and calling it unconditionally would make more sense IMO.

Reply:
Let me try to understand your point: 

- CopyOVSUtilities is called with the condition `if ("$usrBinPath" -ne "$hostUsrBinPath")`. And in this function, we only copy the OVS utilities which are located under path "openvswitch/usr/bin"
- When `$InstallUserspace -eq $true`, copy the OVS userspace binaries `openswitch/usr/sbin` to the host path, then register Windows native services, and configure the OVSDB with OVS version.

Is this what you expect @antoninbas ?

Reply:
I think these lines could be replaced with:
```powershell
    if ($InstallUserspace -eq $true) {
        InstallOVSServices($OVSInstallPath)
    }
    # Ensure that OVS utility programs are available under C:\openvswitch and are in the PATH
    CopyOVSUtilitiesIfNeeded($OVSInstallPath)
```

By the way, does it make sense to hardcode `C:\openvswitch` in `CopyOVSUtilities` instead of using `OVSInstallDir`?